errors/middleware-upgrade-guide.md

```diff
@@ -0,0 +1,309 @@
+# Middleware Upgrade Guide
+
+As we work on improving Middleware for General Availability (GA), we've made some changes to the Middleware APIs (and how you define Middleware in your application) based on your feedback.
+
+This upgrade guide will help you understand the changes and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:
+
+- Currently use the beta Next.js Middleware features
+- Choose to upgrade to the next stable version of Next.js
+
+## Using Next.js Middleware on Vercel
+
+If you're using Next.js on Vercel, your existing deploys using Middleware will continue to work, and you can continue to deploy your site using Middleware. When you upgrade your site to the next stable version of Next.js (`v12.2`), you will need to follow this upgrade guide to update your Middleware.
+
+## Breaking changes
+
+1. [No Nested Middleware](#no-nested-middleware)
+2. [No Response Body](#no-response-body)
+3. [Cookies API Revamped](#cookies-api-revamped)
+4. [No More Page Match Data](#no-more-page-match-data)
+5. [Executing Middleware on Internal Next.js Requests](#executing-middleware-on-internal-nextjs-requests)
+
+## No Nested Middleware
+
+### Summary of changes
+
+- Define a single Middleware file at the root of your project
+- No need to prefix the file with an underscore
+- A custom matcher can be used to define matching routes using an exported config object
+
+### Explanation
+
+Previously, you could create a `_middleware.ts` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created. Beta customers found this route matching confusing. For example:
+
+- Middleware in `pages/dashboard/_middleware.ts`
+- Middleware in `pages/dashboard/users/_middleware.ts`
+- A request to `/dashboard/users/*` **would match both.**
+
+Based on customer feedback, we have replaced this API with a single root Middleware.
+
+### How to upgrade
+
+You should declare **one single Middleware file** in your application, which should be located at the root of the project directory (**not** inside of the `pages` directory), and named **without** an `_` prefix. Your Middleware file can still have either a `.ts` or `.js` extension.
+
+Middleware will be invoked for **every route in the app**, and a custom matcher can be used to define matching filters. The following is an example for a Middleware that triggers for `/about/*`, the custom matcher is defined in an exported config object:
```

Comment:
Below the request will be executing for both `/about/*` and `/dashboard/:path*` but the line only mentions `/about/*` :thonk: